src/packages/emmett-mongodb/src/eventStore/mongoDBEventStore.ts

```diff
@@ -154,89 +210,159 @@ export class MongoDBEventStore implements EventStore<number> {
       });
     }
 
-    // TODO: better error here, should rarely happen if ever
-    // if another error was not thrown before this
+    // TODO: error handling / implement retries
     if (!stream) throw new Error('Failed to create stream');
 
     assertExpectedVersionMatchesCurrent(
-      stream.events.length,
+      stream.metadata.streamPosition,
       options?.expectedStreamVersion,
       MongoDBEventStoreDefaultStreamVersion,
     );
 
+    const updates: UpdateFilter<EventStream> = {
+      $push: { events: { $each: eventCreateInputs } },
+      $set: {
+        'metadata.updatedAt': new Date(),
+        'metadata.streamPosition':
+          stream.metadata.streamPosition + BigInt(events.length),
+      },
+    };
+
+    if (this.asyncProjections) {
```

Comment:
**CONCERN:** Async projections should be handled after events are appended. Without that, you may update async projections and not have event appended. which will cause state inconsistency.

Reply:
Moved to after the events are appended